src/providers/ogr/qgsogrprovider.cpp

```diff
@@ -1132,7 +1128,17 @@ void QgsOgrProviderUtils::setRelevantFields( OGRLayerH ogrLayer, int fieldCount,
         if ( OGRFieldDefnH field = OGR_FD_GetFieldDefn( featDefn, firstAttrIsFid ? i - 1 : i ) )
         {
           const char *fieldName = OGR_Fld_GetNameRef( field );
-          if ( qstrcmp( fieldName, ORIG_OGC_FID ) != 0 )
+          // This is implemented a bit in a hacky way, but in case we are acting on a layer
+          // with a subset filter, do not ignore fields that are found in the
+          // where clause. We do this in a rough way, by looking, in a case
+          // insensitive way, if the current field name is in the subsetString,
+          // so we potentially don't ignore fields we could, in situations like
+          // subsetFilter == "foobar = 2", and there's a "foo" or "bar" field.
+          // Better be safe than sorry.
+          // We could argue that OGR_L_SetIgnoredFields() should be aware of
```

Comment:
s/could/should/

FWIW, it's what QGIS expression based feature requests do -- we silently add any attributes required for client side filtering and sorting to the request.